### PR TITLE
[CH] Optimize aggregate state serialization performance 

### PR DESCRIPTION
**Pull request by liuneng1994**
## What changes were proposed in this pull request?

Optimize aggregate state serialization performance, convert fixed size agg state to fixed string,  
convert variable size aggregation state to string.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by baibaichen** [comment]
--conf spark.sql.autoBroadcastJoinThreshold=100MB
HDSF
TPCH 100

image

File: cpp-ch/local-engine/Builder/SerializedPlanBuilder.cpp [hunk 1]
```diff
@@ -259,7 +259,7 @@ std::shared_ptr<substrait::Type> SerializedPlanBuilder::buildType(const DB::Data
         res->mutable_i32()->set_nullability(type_nullability);
     else if (which.isInt64())
         res->mutable_i64()->set_nullability(type_nullability);
-    else if (which.isString() || which.isAggregateFunction())
+    else if (which.isString() || which.isAggregateFunction() || which.isFixedString())
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Builder/SerializedPlanBuilder.cpp:262]: replace `which.isString() || which.isFixedString()` with `which.isStringOrFixedString()`

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+namespace local_engine
+{
+DB::ColumnWithTypeAndName convertAggregateStateToFixedString(DB::ColumnWithTypeAndName col)
+{
+    if (!isAggregateFunction(col.type))
+    {
+        return col;
+    }
+    const auto *aggregate_col = checkAndGetColumn<ColumnAggregateFunction>(*col.column);
+    size_t state_size = aggregate_col->getAggregateFunction()->sizeOfData();
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:36]: Is the state size of max(String) and collect_list(x) fixed?

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    auto res_type = std::make_shared<DataTypeFixedString>(state_size);
+    auto res_col = res_type->createColumn();
+    res_col->reserve(aggregate_col->size());
+    for (const auto & item : aggregate_col->getData())
+    {
+        res_col->insertData(item, state_size);
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:42]: I think we could use `memcpy` here to copy the whole PODArray into another direclty.

File: cpp-ch/local-engine/Storages/IO/NativeWriter.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    settings.low_cardinality_max_dictionary_size = 0;
+
+    ISerialization::SerializeBinaryBulkStatePtr state;
+    serialization.serializeBinaryBulkStatePrefix(*full_column, settings, state);
+    serialization.serializeBinaryBulkWithMultipleStreams(*full_column, offset, limit, settings, state);
+    serialization.serializeBinaryBulkStateSuffix(settings, state);
+}
+
+size_t NativeWriter::write(const DB::Block & block)
+{
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/IO/NativeWriter.cpp:51]: what is the purpose of `IndexOfBlockForNativeFormat` in the original `NativeWriter`

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 2 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:36]: exclude collect_list, max(String) will use sort aggregate and fallback now

File: cpp-ch/local-engine/Storages/IO/NativeWriter.cpp (hunk 4 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/NativeWriter.cpp:53]: some index data..useless for us

File: cpp-ch/local-engine/Storages/IO/NativeWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+void NativeWriter::flush()
+{
+    ostr.next();
+}
+
+static void writeData(const ISerialization & serialization, const ColumnPtr & column, WriteBuffer & ostr, UInt64 offset, UInt64 limit)
+{
+    /** If there are columns-constants - then we materialize them.
+      * (Since the data type does not know how to serialize / deserialize constants.)
+      */
```
Comment (zhanglistar) [inline cpp-ch/local-engine/Storages/IO/NativeWriter.cpp:38]: Maybe we can design some protocol to write less data for constant column to reducing IO amount.

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+using namespace DB;
+
+namespace local_engine
+{
+
+bool isFixedSizeStateAggregateFunction(const String& name)
+{
+    // TODO max(String) should exclude, but fallback now
+    static const std::set<String> function_set = {"min", "max", "sum", "count", "avg"};
```
Comment (zhanglistar) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:34]: According https://spark.apache.org/docs/latest/sql-ref-functions-builtin.html, maybe other functions can also be added to the set, like mean.

Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:34]: 这里我会换一个方案来同时支持定长和变长的聚合

Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:35]: 目前先用这个，后面遇到了在更新

File: gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala [hunk 7]
```diff
@@ -438,8 +438,10 @@ class GlutenAdaptiveQueryExecSuite extends AdaptiveQueryExecSuite with GlutenSQL
   test("gluten Exchange reuse") {
     withSQLConf(
       SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
-      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "100",
-      SQLConf.SHUFFLE_PARTITIONS.key -> "5") {
+      // magic threshold, ch backend has two bhj when threshold is 100
```
Comment (zzcclp) [inline gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala:441]: @PHILO-HE @rui-mo please help to check, thanks.

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala [hunk 8]
```diff
@@ -438,8 +438,10 @@ class GlutenAdaptiveQueryExecSuite extends AdaptiveQueryExecSuite with GlutenSQL
   test("gluten Exchange reuse") {
     withSQLConf(
       SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
-      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "100",
-      SQLConf.SHUFFLE_PARTITIONS.key -> "5") {
+      // magic threshold, ch backend has two bhj when threshold is 100
```
Comment (zzcclp) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala:441]: @PHILO-HE @rui-mo please help to check, thanks.

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    const auto *aggregate_col = checkAndGetColumn<ColumnAggregateFunction>(*col.column);
+    // only support known fixed size aggregate function
+    if (!isFixedSizeAggregateFunction(aggregate_col->getAggregateFunction()))
+    {
+        return col;
+    }
+    size_t state_size = aggregate_col->getAggregateFunction()->sizeOfData();
+    auto res_type = std::make_shared<DataTypeFixedString>(state_size);
+    auto res_col = res_type->createColumn();
+    PaddedPODArray<UInt8> & column_chars_t = assert_cast<ColumnFixedString &>(*res_col).getChars();
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:64]: use `ColumnFixedString::reserve` is better

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    column_chars_t.reserve(aggregate_col->size() * state_size);
+    for (const auto & item : aggregate_col->getData())
+    {
+        column_chars_t.insert_assume_reserved(item, item + state_size);
+    }
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:69]: for small state objects, could we try to use `memcpy` ?

too much function call may have some performance issue

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala (hunk 8 above)
Comment (PHILO-HE) [reply gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala:441]: As velox backend CI passed, I think this change is OK. Thanks!

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 9 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:64]: 同下

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 10 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:69]: 这个写法是参考arrowColumnToCHColumn的readColumnWithNumericData。应该不会有性能问题

File: cpp-ch/local-engine/Storages/IO/NativeWriter.cpp (hunk 5 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/NativeWriter.cpp:43]: 这个是CH copy过来的。。暂时不做优化

**Comment by liuneng1994** [comment]
Run Gluten Clickhouse CI

**Comment by lgbo-ustc** [comment]
LGTM

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+{
+    return data_types.front()->isValueRepresentedByNumber();
+}
+
+bool isFixedSizeAggregateFunction(DB::AggregateFunctionPtr function)
+{
+    return isFixedSizeStateAggregateFunction(function->getName()) && isFixedSizeArguments(function->getArgumentTypes());
+}
+
+DB::ColumnWithTypeAndName convertAggregateStateToFixedString(DB::ColumnWithTypeAndName col)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:49]: 尽量避免传值，避免不必要的复制开销。

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        agg_function->create(place);
+
+        auto value = col.column->getDataAt(i);
+        memcpy(place, value.data, value.size);
+
+        vec.push_back(place);
+    }
+    return DB::ColumnWithTypeAndName(std::move(res_col), type, col.name);
+}
+DB::Block convertAggregateStateInBlock(DB::Block block)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:123]: 避免传值

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+bool isFixedSizeStateAggregateFunction(const String& name)
+{
+    static const std::set<String> function_set = {"min", "max", "sum", "count", "avg"};
+    return function_set.contains(name);
+}
+
+bool isFixedSizeArguments(DataTypes data_types)
+{
+    return data_types.front()->isValueRepresentedByNumber();
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:41]: isValueUnambiguouslyRepresentedInFixedSizeContiguousMemoryRegion 
使用这个接口是否更为准确。

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        auto value = col.column->getDataAt(i);
+        memcpy(place, value.data, value.size);
+
+        vec.push_back(place);
+    }
+    return DB::ColumnWithTypeAndName(std::move(res_col), type, col.name);
+}
+DB::Block convertAggregateStateInBlock(DB::Block block)
+{
+    ColumnsWithTypeAndName columns;
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:125]: 加上reserve

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    static const std::set<String> function_set = {"min", "max", "sum", "count", "avg"};
+    return function_set.contains(name);
+}
+
+bool isFixedSizeArguments(DataTypes data_types)
+{
+    return data_types.front()->isValueRepresentedByNumber();
+}
+
+bool isFixedSizeAggregateFunction(DB::AggregateFunctionPtr function)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:44]: 避免传值，其他类似地方也可以优化下。

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+{
+    return isFixedSizeStateAggregateFunction(function->getName()) && isFixedSizeArguments(function->getArgumentTypes());
+}
+
+DB::ColumnWithTypeAndName convertAggregateStateToFixedString(DB::ColumnWithTypeAndName col)
+{
+    if (!WhichDataType(col.type).isAggregateFunction())
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:51]: 这里的判断可以去掉，改成下面判断aggregate_col是否为nullptr

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        return col;
+    }
+    const auto *aggregate_col = checkAndGetColumn<ColumnAggregateFunction>(*col.column);
+    auto res_type = std::make_shared<DataTypeString>();
+    auto res_col = res_type->createColumn();
+    PaddedPODArray<UInt8> & column_chars = assert_cast<ColumnString &>(*res_col).getChars();
+    column_chars.reserve(aggregate_col->size() * 60);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:83]: 60这个数字是怎么来的？ 会不会导致分配超出实际需要的内存

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 16 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:51]: 这种类型AggregateFunction(sum, Nullable(Int64)) 是否可以走fixed string

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 13 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:41]: 先限制在数字类型，其他类型没测试过

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 17 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:83]: 测试代码，删了

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 16 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:51]: bug, 需要remove nullable

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 18]
```diff
@@ -36,23 +36,23 @@ bool isFixedSizeStateAggregateFunction(const String& name)
...
 
-bool isFixedSizeArguments(DataTypes data_types)
+bool isFixedSizeArguments(const DataTypes& data_types)
 {
-    return data_types.front()->isValueRepresentedByNumber();
+    return removeNullable(data_types.front())->isValueRepresentedByNumber();
 }
 
-bool isFixedSizeAggregateFunction(DB::AggregateFunctionPtr function)
+bool isFixedSizeAggregateFunction(const DB::AggregateFunctionPtr& function)
```
Comment (zhanglistar) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:44]: shared_ptr
不用传引用吧

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp [hunk 19]
```diff
@@ -58,6 +60,30 @@ DB::ColumnWithTypeAndName convertAggregateStateToFixedString(DB::ColumnWithTypeA
...
+    auto res_col = res_type->createColumn();
+    PaddedPODArray<UInt8> & column_chars = assert_cast<ColumnString &>(*res_col).getChars();
+    column_chars.reserve(aggregate_col->size() * 60);
+    IColumn::Offsets & column_offsets = assert_cast<ColumnString &>(*res_col).getOffsets();
+    auto value_writer = WriteBufferFromVector<PaddedPODArray<UInt8>>(column_chars);
+    column_offsets.reserve(aggregate_col->size());
+    for (const auto & item : aggregate_col->getData())
+    {
+        aggregate_col->getAggregateFunction()->serialize(item, value_writer);
+        writeChar('\0', value_writer);
```
Comment (zhanglistar) [inline cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:89]: 这个有必要？

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 18 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:44]: 就这样吧，没有太大区别

File: cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp (hunk 19 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/IO/AggregateSerializationUtils.cpp:89]: 字符串需要\0结尾，必须的

File: gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala (hunk 7 above)
Comment (rui-mo) [reply gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/adaptive/GlutenAdaptiveQueryExecSuite.scala:440]: LGTM.

**Comment by lgbo-ustc** [comment]
LGTM

**Comment by taiyang-li** [comment]
LGTM

**Review by baibaichen — APPROVED** [review]
cool, let's merge

**Pull request merged**